### PR TITLE
3.4 modifier protocols

### PR DESCRIPTION
**Pull request by andrewkerr9000**
Functionally complete, still cleaning up, may be annoying edge cases.

TODO

* ~~Remove redundant supported modifier protocols parameter from handshake server~~
* ~~Consider moving supported application protocols into app protocol repo~~
* ~~Consider model for modifier protocols (ID & version number not ideal)~~
* ~~Ensure server won't accept a modifier protocol excluded from configured supported modifiers~~
* ~~Test for no intersection between client & server supported protocols-~~
* ~~Parameterised test for combinations of modifiers applied to a real netty channel~~
* ~~How to document available modifier protocols?~~
* ~~still some TODOs left in code~~
* ~~refactor methods using supported protocols into protocol repository~~
* ~~support alternate compression methods~~

### CODE REVIEW

**Comment by andrewkerr9000** [comment]
gzip, lz4 (with mvn dependency) added. Variants of gzip and lz4, treated as separate algorithms (but sharing client or server installers). No additional parameters for algorithm are configurable, e.g. it may make sense to configure deflate level for gzip.

**Comment by andrewkerr9000** [comment]
test failures are unrelated.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/CausalClusteringSettings.java [hunk 1]
```diff
@@ -490,4 +494,16 @@
...
     public static final Setting<String> ssl_policy =
             prefixSetting( "causal_clustering.ssl_policy", STRING, NO_DEFAULT );
+
+    @Description( "Raft protocol implementation versions that this instance will allow in negotiation as a comma-separated list." +
+            " Order is not relevant: the greatest value will be preferred. An empty list will allow all supported versions" )
+    public static final Setting<List<Integer>> raft_implementations =
+            setting( "causal_clustering.protocol_implementations.raft", list( ",", INTEGER ), "" );
+
+    @Description( "Network compression algorithms that this instance will allow in negotiation as a comma-separated list." +
+            " Listed in descending order of preference for incoming connections. An empty list implies no compression." +
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/CausalClusteringSettings.java:504]: Perhaps clarify with: " For outgoing connections this merely specifies the allowed set of algorithms and the preference of the remote peer will be used for making the decision."

(assuming that is correct)

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/protocol/handshake/SupportedProtocols.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.neo4j.causalclustering.protocol.Protocol;
+
+public abstract class SupportedProtocols<U extends Comparable<U>,T extends Protocol<U>>
+{
+    private final Protocol.Category<T> category;
+    private final List<U> versions;
+
+    /**
+     * @param versions Empty means support everything
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/protocol/handshake/SupportedProtocols.java:35]: Would be nice to have a complete javadoc.

**Review by martinfurmanski — APPROVED** [review]
Would be nice with a bit more Javadoc around SupportedProtocols and related to guide one as to how it all fits together. The names are good, but a bit of text would help.

Would also be good to have an IT based on ClusterRule which replicates data around using some compression scheme or perhaps even all of them as a parameterised test.

Can be merged as-is though and followed up in a subsequent PR to speed up the process.

**Pull request merged**